### PR TITLE
Fix status check in UploadResponse object

### PR DESCRIPTION
**Pull request by KineticCookie**
* Introduced additional fields to the Model object
* Fixed Status enums
* Changed status poll

### CODE REVIEW

File: hydrosdk/model.py [hunk 1]
```diff
@@ -561,91 +564,67 @@ def __init__(self, name, id_, contract, version, metadata):
...
-        logger = logging.getLogger("ModelDeploy")
-        try:
-            url = "/api/v2/model/version/{}/logs".format(self.model_version_id)
-            logs_response = self.model.cluster.request("GET", url, stream=True)
-            self._logs_iterator = sseclient.SSEClient(logs_response).events()
-        except RuntimeError:
-            logger.exception("Unable to get build logs")
-            self._logs_iterator = None
-        return self._logs_iterator
+        url = "/api/v2/model/version/{}/logs".format(self.model_version_id)
```
Comment (trthhrtz) [inline hydrosdk/model.py:583]: Why no exception handling?

File: hydrosdk/model.py [hunk 2]
```diff
@@ -561,91 +564,67 @@ def __init__(self, name, id_, contract, version, metadata):
...
-                self.last_log = next(self._logs_iterator).data
-                self._status = BuildStatus.BUILDING
-        except StopIteration:
-            if not self._status == BuildStatus.FINISHED:
-                self._status = BuildStatus.FAILED
+        model = Model.find(self.cluster, self.model.name, self.model.version)
+        if model:
+            self.model = model
+        else:
+            raise ValueError(
```
Comment (trthhrtz) [inline hydrosdk/model.py:597]: There is an exception handling inside Model.find

File: hydrosdk/model.py [hunk 3]
```diff
@@ -561,91 +564,67 @@ def __init__(self, name, id_, contract, version, metadata):
...
-        except RuntimeError:
-            logger.exception("Unable to get build logs")
-            self._logs_iterator = None
-        return self._logs_iterator
+        url = "/api/v2/model/version/{}/logs".format(self.model_version_id)
+        logs_response = self.model.cluster.request("GET", url, stream=True)
+        return sseclient.SSEClient(logs_response).events()
 
-    def set_status(self) -> None:
+    def poll_model(self) -> None:
```
Comment (trthhrtz) [inline hydrosdk/model.py:587]: New realization of logging is much more cumbersome, after each status check request is sent, then object is deserialized and only then status is checked. But I guess  it is required by the task

File: hydrosdk/model.py (hunk 1 above)
Comment (KineticCookie) [reply hydrosdk/model.py:583]: Well, we can't correctly handle the exception of logs missing: logging the issue just hides the exception and make debugging harder.

If there is a case that we can't get logs, I think it's better to propagate the underlying exception and let user choose the best course of action

File: hydrosdk/model.py (hunk 3 above)
Comment (KineticCookie) [reply hydrosdk/model.py:587]: But it's the only correct option here. Parsing log strings for a "Successful" substring doesn't mean that model finished building.

File: hydrosdk/model.py [hunk 4]
```diff
@@ -477,6 +478,8 @@ def __init__(self, id, name, version, contract, runtime, image, cluster, metadat
         self.version = version
         self.image = image
 
+        self.status = status
```
[the remark this answers is not in the document]
Comment (KineticCookie) [reply hydrosdk/model.py:481]: 👌

File: hydrosdk/model.py [hunk 5]
```diff
@@ -397,6 +378,38 @@ def from_proto(proto, cluster):
             cluster=cluster
         )
 
+    @staticmethod
+    def from_json(cluster, model_json):
```
[the remark this answers is not in the document]
Comment (KineticCookie) [reply hydrosdk/model.py:382]: 👌

File: hydrosdk/model.py (hunk 2 above)
Comment (KineticCookie) [reply hydrosdk/model.py:597]: my bad didn't notice

File: hydrosdk/model.py [hunk 6]
```diff
@@ -561,91 +549,61 @@ def __init__(self, name, id_, contract, version, metadata):
...
+
+    Check the status of the assembly using `get_status()`
+    Check the build logs using `logs()`
     """
-    def __init__(self, model, version_id):
+
+    def __init__(self, model):
         self.cluster = model.cluster
         self.model = model
-        self.model_version_id = version_id
```
Comment (trthhrtz) [inline hydrosdk/model.py:580]: https://github.com/Hydrospheredata/hydro-serving-sdk/blob/1fa3ce331e908cb5ba393e6c3199a8a11663ffd5/hydrosdk/model.py#L279

requires model_version_id

File: hydrosdk/model.py [hunk 7]
```diff
@@ -366,35 +350,47 @@ def find_by_id(cluster, model_id):
...
+        # TODO: get its own endpoint for external model
+        if not model_json.get("runtime"):
+            model_json["runtime"] = {}
+
+        model_runtime = DockerImage(model_json["runtime"].get("name"), model_json["runtime"].get("tag"),
+                                    model_json["runtime"].get("sha256"))
+        model_image = model_json.get("image")
+        model_cluster = cluster
+
+        status = ModelStatus[model_json['status']]
```
Comment (trthhrtz) [inline hydrosdk/model.py:381]: Request to `/api/v2/model/version/{}/{}` returns this kind of response
```
{'model': {'id': 5, 'name': 'ext-model-test'}, 'modelContract': {'modelName': '', 'predict': {'signatureName': 'infer', 'inputs': [{'profile': 'NONE', 'dtype': 'DT_DOUBLE', 'name': 'in1', 'shape': {'dim': [{'size': -1, 'name': ''}, {'size': 2, 'name': ''}], 'unknownRank': False}}], 'outputs': [{'profile': 'NONE', 'dtype': 'DT_DOUBLE', 'name': 'out1', 'shape': {'dim': [{'size': -1, 'name': ''}], 'unknownRank': False}}]}}, 'id': 10, 'metadata': {'additionalProp1': 'prop'}, 'kind': 'External', 'modelVersion': 3, 'created': '2020-04-13T06:11:16.506Z', 'runtime': {}}
```
There is no status

File: hydrosdk/model.py [hunk 8]
```diff
@@ -366,35 +349,50 @@ def find_by_id(cluster, model_id):
...
-            image=DockerImage(name=proto.image.name, tag=proto.image.tag, sha256=proto.image_sha),
-            cluster=cluster
+    def from_json(cluster, model_json):
+        """
+        Internal method used for deserealization of a Model from json object
+        :param cluster:
+        :param model_json: a dictionary
+        :return: A Model instance
+        """
+        if model_json['isExternal']:
```
Comment (trthhrtz) [inline hydrosdk/model.py:365]:
```
 KeyError: 'isExternal'

hydrosdk/model.py:367: KeyError
--------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------
{'model': {'id': 3, 'name': 'linear_regression_prod'}, 'image': {'name': 'linear_regression_prod', 'tag': '2'}, 'modelContract': {'modelName': '', 'predict': {'signatureName': 'infer', 'inputs': [{'profile': 'NONE', 'dtype': 'DT_DOUBLE', 'name': 'in1', 'shape': {'dim': [{'size': -1, 'name': ''}, {'size': 2, 'name': ''}], 'unknownRank': False}}], 'outputs': [{'profile': 'NONE', 'dtype': 'DT_DOUBLE', 'name': 'out1', 'shape': {'dim': [{'size': -1, 'name': ''}], 'unknownRank': False}}]}}, 'id': 33, 'status': 'Assembling', 'metadata': {}, 'modelVersion': 2, 'runtime': {'name': 'hydrosphere/serving-runtime-python-3.6', 'tag': '2.1.0'}, 'created': '2020-04-14T11:46:40.695Z'}
```

File: hydrosdk/model.py [hunk 9]
```diff
@@ -84,9 +84,18 @@ def read_yaml(path):
...
-def read_py(path):
-    pass
+def parse_model_from_json_dict(cluster, json_dict):
+    """
+    Parses json dict which could contain ExternalModel or Model
+    :param cluster:
+    :param json_dict:
+    :return:
+    """
+    is_external = json_dict.get("isExternal", True)
```
Comment (trthhrtz) [inline hydrosdk/model.py:94]:
```
json_dict = {1: 1}
is_external = json_dict.get("isExternal", True)
if is_external:
  print(1)
else:
  print(2)
```
Prints 1

Comment (KineticCookie) [reply hydrosdk/model.py]: That's not a problem, per se. Backend has inconsisted output, depending on the endpoint. With the latest fixes, I tried to make them similar, but there will be situations, when model object is returned without `isExternal` flag. Thus we need some kind of default value.

`True` was a bad option here, I think `False` will be better.

Comment (trthhrtz) [reply hydrosdk/model.py]: @KineticCookie ah, ok. I thought that was unexpected behavior and therefore notified you. In tests some models (LocalModel) were uploaded as external ones though they were uploaded as usual and I thought they were supposed to be class Model in the result.

File: hydrosdk/model.py [hunk 10]
```diff
@@ -366,35 +361,48 @@ def find_by_id(cluster, model_id):
...
+        # TODO: get its own endpoint for external model
+        if not model_json.get("runtime"):
+            model_json["runtime"] = {}
+
+        model_runtime = DockerImage(model_json["runtime"].get("name"), model_json["runtime"].get("tag"),
+                                    model_json["runtime"].get("sha256"))
+        model_image = model_json.get("image")
+        model_cluster = cluster
+
+        status = ModelStatus[model_json['status']]
```
Comment (trthhrtz) [inline hydrosdk/model.py:393]:
```_____________________________________________________________________ test_external_model_find_by_name _____________________________________________________________________

    def test_external_model_find_by_name():
        cluster = get_cluster()
    
        name, contract, metadata = get_ext_model_fields()
    
        created_model = ExternalModel.create(cluster=cluster, name=name, contract=contract, metadata=metadata)
        found_model = ExternalModel.find_by_name(cluster=cluster, name=created_model.name,
>                                                version=created_model.version)

tests/test_model.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
hydrosdk/model.py:538: in find_by_name
    found_model = Model.find(cluster=cluster, name=name, version=version)
hydrosdk/model.py:342: in find
    return parse_model_from_json_dict(cluster, model_json)
hydrosdk/model.py:101: in parse_model_from_json_dict
    return Model.from_json(cluster, json_dict)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cluster = <hydrosdk.cluster.Cluster object at 0x10dca4710>
model_json = {'created': '2020-04-17T06:23:40.627Z', 'id': 49, 'kind': 'External', 'metadata': {'additionalProp1': 'prop'}, ...}

    @staticmethod
    def from_json(cluster, model_json):
        """
        Internal method used for deserealization of a Model from json object
        :param cluster:
        :param model_json: a dictionary
        :return: A Model instance
        """
    
        model_id = model_json["id"]
        model_name = model_json["model"]["name"]
        model_version = model_json["modelVersion"]
        model_contract = contract_from_dict(model_json["modelContract"])
    
        # external model deserialization handling
        # TODO: get its own endpoint for external model
        if not model_json.get("runtime"):
            model_json["runtime"] = {}
    
        model_runtime = DockerImage(model_json["runtime"].get("name"), model_json["runtime"].get("tag"),
                                    model_json["runtime"].get("sha256"))
        model_image = model_json.get("image")
        model_cluster = cluster
    
>       status = ModelStatus[model_json['status']]
E       KeyError: 'status'

hydrosdk/model.py:393: KeyError
```

Should we add for now  the following work-around?
```
        if not model_json.get("status"):
            status = None
        else:
            status = ModelStatus[model_json['status']]
```

**Comment by trthhrtz** [comment]
good to go @KineticCookie :+1:

**Pull request merged**